Objects/complexobject.c

```diff
@@ -474,83 +579,87 @@ complex_hash(PyComplexObject *v)
 
-static PyObject *
-complex_mul(PyObject *v, PyObject *w)
-{
-    Py_complex result;
-    Py_complex a, b;
-    TO_COMPLEX(v, a);
-    TO_COMPLEX(w, b);
-    result = _Py_c_prod(a, b);
-    return PyComplex_FromCComplex(result);
-}
+   Let us consider the addition as an example, assuming that ints are implicitly
+   converted to floats.  We have the following rules (up to variants with changed
+   order of operands):
 
-static PyObject *
-complex_div(PyObject *v, PyObject *w)
-{
-    Py_complex quot;
-    Py_complex a, b;
-    TO_COMPLEX(v, a);
-    TO_COMPLEX(w, b);
-    errno = 0;
-    quot = _Py_c_quot(a, b);
-    if (errno == EDOM) {
-        PyErr_SetString(PyExc_ZeroDivisionError, "division by zero");
-        return NULL;
-    }
-    return PyComplex_FromCComplex(quot);
-}
+       complex(x, y) + complex(u, v) = complex(x + u, y + v)
+       float(x) + complex(u, v) = complex(x + u, v)
+
+   Similar rules are implemented for subtraction, multiplication and division.
+   See C11's Annex G, sections G.5.1 and G.5.2.
+ */
+
+#define COMPLEX_BINOP(name, func)                             \
+    static PyObject *                                         \
+    complex_##name(PyObject *v, PyObject *w)                  \
+    {                                                         \
+        Py_complex a;                                         \
+        errno = 0;                                            \
+        if (PyComplex_Check(w)) {                             \
+            Py_complex b = ((PyComplexObject *)(w))->cval;    \
+            if (PyComplex_Check(v)) {                         \
+                a = ((PyComplexObject *)(v))->cval;           \
+                a = _Py_c_##func(a, b);                       \
+            }                                                 \
+            else if (real_to_double(&v, &a.real) < 0) {       \
+                return v;                                     \
+            }                                                 \
+            else {                                            \
+                a = _Py_rc_##func(a.real, b);                 \
+            }                                                 \
+        }                                                     \
+        else {                                                \
+            a = ((PyComplexObject *)(v))->cval;               \
+            double b;                                         \
+            if (real_to_double(&w, &b) < 0) {                 \
+                return w;                                     \
+            }                                                 \
+            a = _Py_cr_##func(a, b);                          \
+        }                                                     \
+        if (errno == EDOM) {                                  \
+            PyErr_SetString(PyExc_ZeroDivisionError,          \
+                            "division by zero");              \
```

Comment:
I know that `real_to_double` may set `w` to `Py_NewRef(Py_NotImplemented)` and it then becomes immortal. My question is: don't we have leaks because of this? at which point is the original `w` decrefed in case of an error?

Reply:
> at which point is the original w decrefed in case of an error?

That happens in binary_op1:
https://github.com/python/cpython/blob/8c22eba877225ab29cd64672dcb97f09a5f7336f/Objects/abstract.c#L952-L976